Fixing parse error returning 500  instead of 400.

cheftako opened the pull request:
**What this PR does / why we need it**:
Added code to labels to allow it to return StatusError.
Moved helper code out of meta/v1 to selectors to prevent circular dep.
metav1 -> labels -> errors -> metav1 introduced by above change.
Added ErrorToBadRequest method and relevant tests
ErrorToBadRequest upgrades a simple error to a BadRequest StatusError.
If passed a BadRequest it will just pass it through.
If it is passed any other type of StatusError it will panic.
Fixed the selector to return StatusError and not a generic error.
Got all the main apimachinery field selector cases changed over.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #39730

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Invalid field labels and values are returning 500 series errors. They should now be returning 400's.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40773)
<!-- Reviewable:end -->

deads2k commented on staging/src/k8s.io/client-go/pkg/api/conversion.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime"
+	selectors "k8s.io/apimachinery/pkg/selectors"
```
Where did this package come from?

cheftako replied:
I added it to break a circular dependency I created when I had labels depend on StatusError.

cheftako committed:
Fixing parse error returning 500  tinstead of 400.

    Added code to labels to allow it to return StatusError.
    Moved helper code out of meta/v1 to selectors to prevent circular dep.
    metav1 -> labels -> errors -> metav1 introduced by above change.
    Added ErrorToBadRequest method and relevant tests
    ErrorToBadRequest upgrades a simple error to a BadRequest StatusError.
    If passed a BadRequest it will just pass it through.
    Fixed the selector to return StatusError and not a generic error.
    Got all the main apimachinery field selector cases changed over.
    Minimal client-go changes to get things to compile with apimachinery
    changes. fixes #39730

cheftako commented:
@k8s-bot non-cri node e2e test this

cheftako commented:
@k8s-bot verify test this

smarterclayton commented:
What's the context for this change?  Why is this necessary?

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/fields/selector.go:
```diff
@@ -77,7 +78,7 @@ func (t *hasTerm) RequiresExactMatch(field string) (value string, found bool) {
 	field, value, err := fn(t.field, t.value)
 	if err != nil {
-		return nil, err
+		return nil, errors.ErrorToBadRequest(err)
```
This is not what I would expected.  Generic field selectors have nothing to do with API errors.  What are you trying to accomplish here?

cheftako replied:
I'm trying to improve the quality of our response codes so that they reflect what is going wrong. Currently we always return a 500 no matter what went wrong. My impression of the errors coming out of the fields and labels packages is that they indicate that the pieces passed in be the client (value or label) are bad and will never work with this version of the API server. In that case 400 is the more appropriate HTTP error.

cheftako commented:
> What's the context for this change? Why is this necessary?

https://github.com/kubernetes/kubernetes/issues/39730

smarterclayton commented:
Normally we would do that at a higher level - labels and fields in general are not API specific, so we would return errors outside of it.   A couple of options:

1. type the errors in the conversion (conversion is allowed to know about transition from API to internal, which is a close enough context)
2. Use typed errors in fields and errors and expose a public test method (IsLabelError) and add that public test method to the api server where it does error conversion for return.

smarterclayton commented:
The gymnastics required to break the circular dependency in this case are due to the intent that API not expose a significant amount of detail.

Also, option:

3. Add a new interface type "FieldValidationError" that has a method "ValidationErrors" that returns `[]pkg/util/validation/field.ValidationError`, and implement that method on the errors in labels / fields, then check for that interface type in the API server (as a simpler version of Status()).

Of all of them I prefer 1, because that's the only place that knows which field carried the error, and in many cases this is a field processing error.  We would probably want to return 422 invalid when this is a selector embedded in an object, not just 400.

cheftako commented:
@k8s-bot cvm gke e2e test this

cheftako commented:
@k8s-bot non-cri node e2e test this

crassirostris commented:
@cheftako @smarterclayton How this PR is doing? Is it gonna make it in 1.6 or in a patch release?

smarterclayton commented:
Not sure that the current state of the PR matches what I expect.  I don't expect selectors to move out of metav1.

crassirostris commented:
@smarterclayton Thanks for the update!

OK, gotcha. What about this problem: https://github.com/kubernetes/kubernetes/issues/39730 ?

sttts commented:
If this is still relevant, please split it into one commit with the pure mechanical move and one with the actual change. It's hard to review this way.
